File: backend/app/services/rhythm.py
"""Pure rhythm math: pattern validation and expected-onset computation.

All times are in milliseconds. Taps are submitted relative to the first
count-in click; the count-in is one full measure, so expected onsets are
offset by count_in_ms.
"""

DURATION_BEATS = {"w": 4.0, "h": 2.0, "q": 1.0, "8": 0.5, "16": 0.25}


def parse_time_signature(time_signature: str) -> tuple[int, int]:
    try:
        num, denom = time_signature.split("/")
        return int(num), int(denom)
    except (ValueError, AttributeError):
        raise ValueError(f"Invalid time signature: {time_signature!r}")


def beats_per_measure(time_signature: str) -> float:
    """Beats per measure expressed in quarter notes."""
    num, denom = parse_time_signature(time_signature)
    return num * (4.0 / denom)


def event_beats(event: dict) -> float:
    duration = event.get("duration")
    if duration not in DURATION_BEATS:
        raise ValueError(f"Invalid duration: {duration!r}")
    beats = DURATION_BEATS[duration]
    if event.get("dots", 0) == 1:
        beats *= 1.5
    return beats


def validate_pattern(pattern: dict, time_signature: str, num_measures: int) -> None:
    """Raise ValueError if the pattern's total beats don't fill the measures exactly."""
    events = pattern.get("events", [])
    if not events:
        raise ValueError("Pattern must contain at least one event")
    total = sum(event_beats(e) for e in events)
    expected = beats_per_measure(time_signature) * num_measures
    if abs(total - expected) > 1e-6:
        raise ValueError(
            f"Pattern total of {total} beats does not fill {num_measures} measures "
            f"of {time_signature} ({expected} beats)"
        )
    if not any(e.get("type") == "note" for e in events):
        raise ValueError("Pattern must contain at least one note")
    last = events[-1]
    if last.get("tieToNext"):
        raise ValueError("Last event cannot be tied to a following note")
    for i, e in enumerate(events):
        if e.get("tieToNext"):
            if e.get("type") != "note":
                raise ValueError("Only notes can be tied")
            if events[i + 1].get("type") != "note":
                raise ValueError("A tie must connect two notes")


def beat_ms(tempo_bpm: int) -> float:
    return 60000.0 / tempo_bpm


def count_in_ms(time_signature: str, tempo_bpm: int) -> float:
    return beats_per_measure(time_signature) * beat_ms(tempo_bpm)


def expected_onsets(pattern: dict, time_signature: str, tempo_bpm: int) -> list[float]:
    """Onset times (ms, relative to the first count-in click) the user must tap.

    Rests and tied-into notes consume time but produce no onset.
    """
    bms = beat_ms(tempo_bpm)
    t = count_in_ms(time_signature, tempo_bpm)
    onsets: list[float] = []
    tied_into = False
    for event in pattern.get("events", []):
        if event.get("type") == "note" and not tied_into:
            onsets.append(t)
        t += event_beats(event) * bms
        tied_into = bool(event.get("tieToNext")) and event.get("type") == "note"
    return onsets


def total_duration_ms(pattern: dict, time_signature: str, tempo_bpm: int) -> float:
    """Length of the exercise (count-in + all events), in ms."""
    bms = beat_ms(tempo_bpm)
    events_ms = sum(event_beats(e) for e in pattern.get("events", [])) * bms
    return count_in_ms(time_signature, tempo_bpm) + events_ms

File: frontend/src/lib/rhythm.ts
import type { Pattern, PatternEvent } from '../api/types'

export const DURATION_BEATS: Record<string, number> = {
  w: 4,
  h: 2,
  q: 1,
  '8': 0.5,
  '16': 0.25,
}

export function eventBeats(event: PatternEvent): number {
  let beats = DURATION_BEATS[event.duration] ?? 0
  if (event.dots === 1) {
    beats *= 1.5
  }
  return beats
}

export function beatsPerMeasure(timeSignature: string): number {
  const [numerator, denominator] = timeSignature.split('/').map(Number)
  if (!numerator || !denominator) return 4
  return numerator * (4 / denominator)
}

export function beatMs(tempoBpm: number): number {
  return 60000 / tempoBpm
}

export function countInMs(timeSignature: string, tempoBpm: number): number {
  return beatsPerMeasure(timeSignature) * beatMs(tempoBpm)
}

/**
 * Onset times in ms, relative to the first count-in click, for every event the
 * user must tap. Rests and tied-into notes consume time but produce no onset.
 */
export function expectedOnsets(pattern: Pattern, timeSignature: string, tempoBpm: number): number[] {
  const bms = beatMs(tempoBpm)
  let t = countInMs(timeSignature, tempoBpm)
  const onsets: number[] = []
  let tiedInto = false
  for (const event of pattern.events) {
    if (event.type === 'note' && !tiedInto) {
      onsets.push(t)
    }
    t += eventBeats(event) * bms
    tiedInto = Boolean(event.tieToNext) && event.type === 'note'
  }
  return onsets
}

/**
 * For each onset (in order), the index of the pattern event it belongs to.
 * Used to map per-onset scoring results back onto rendered notes.
 */
export function onsetEventIndices(pattern: Pattern): number[] {
  const indices: number[] = []
  let tiedInto = false
  pattern.events.forEach((event, index) => {
    if (event.type === 'note' && !tiedInto) {
      indices.push(index)
    }
    tiedInto = Boolean(event.tieToNext) && event.type === 'note'
  })
  return indices
}

/** Length of count-in plus the full pattern, in ms. */
export function totalDurationMs(pattern: Pattern, timeSignature: string, tempoBpm: number): number {
  const bms = beatMs(tempoBpm)
  const patternBeats = pattern.events.reduce((sum, event) => sum + eventBeats(event), 0)
  return countInMs(timeSignature, tempoBpm) + patternBeats * bms
}

/** Splits the events into measures based on the time signature (greedy by beats). */
export function splitIntoMeasures(pattern: Pattern, timeSignature: string): PatternEvent[][] {
  const perMeasure = beatsPerMeasure(timeSignature)
  const measures: PatternEvent[][] = []
  let current: PatternEvent[] = []
  let beats = 0
  for (const event of pattern.events) {
    current.push(event)
    beats += eventBeats(event)
    if (beats >= perMeasure - 1e-6) {
      measures.push(current)
      current = []
      beats = 0
    }
  }
  if (current.length > 0) {
    measures.push(current)
  }
  return measures
}
